src/backend/env/auction.rs

```diff
@@ -0,0 +1,393 @@
+use std::collections::BTreeSet;
+
+use candid::Principal;
+use ic_ledger_types::{
+    AccountIdentifier, Memo, Subaccount, Tokens, DEFAULT_FEE, DEFAULT_SUBACCOUNT,
+};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    env::invoices::{self},
+    id, mutate, read, time,
+};
+
+use super::{
+    invoices::{main_account, principal_to_subaccount},
+    token::Token,
+    user::UserId,
+    State, Time,
+};
+
+pub const AUCTION_ICP_SUBACCOUNT: [u8; 32] = [
+    0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1,
+];
+
+pub fn auction_account() -> AccountIdentifier {
+    AccountIdentifier::new(&id(), &Subaccount(AUCTION_ICP_SUBACCOUNT))
+}
+
+#[derive(Clone, Eq, Debug, Default, PartialEq, Serialize, Deserialize)]
+pub struct Bid {
+    pub user: UserId,
+    pub amount: Token,
+    // e8s for 1 TAGGR cent
+    pub e8s_per_token: u64,
+    timestamp: Time,
+}
+
+impl PartialOrd for Bid {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Bid {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        if self.e8s_per_token != other.e8s_per_token {
+            // prioritize higher bids
+            return self.e8s_per_token.cmp(&other.e8s_per_token);
+        }
+        if self.timestamp != other.timestamp {
+            // prioritize older bids
+            return other.timestamp.cmp(&self.timestamp);
+        }
+        if self.amount != other.amount {
+            // prioritize larger bids
+            return self.amount.cmp(&other.amount);
+        }
+        self.user.cmp(&other.user)
+    }
+}
+
+#[derive(Default, Serialize, Deserialize)]
+pub struct Auction {
+    pub amount: Token,
+    bids: BTreeSet<Bid>,
+}
+
+impl Auction {
+    /// Checks if there is a sell out.
+    pub fn sell_out(&self) -> bool {
+        self.bids.iter().map(|bid| bid.amount).sum::<u64>() >= self.amount
+    }
+
+    /// Returns the highest bids if they form a sell out.
+    pub fn get_bids(&mut self) -> Vec<Bid> {
+        if !self.sell_out() {
+            return Default::default();
+        }
+        let mut amount = self.amount;
+        let mut bids = Vec::default();
+
+        while amount > 0 {
+            let Some(bid) = self.bids.pop_last() else {
```

Comment:
It looks like leftover bids are not cleared for the next auction (which seems intentional given the re-insertion of a remainder bid below).

If I've understood this correctly, as long as there is at least one *disinterested* but rational user on TAGGR with some minuscule amount of ICP that they're willing to part with, it makes sense for them to submit a bit for the maximum amount of TAGGR for the minimum possible ICP price.

If there's ever a scenario where the auction would otherwise be cancelled due to not enough bids, what would actually happen is that this described user would receive all of the auctioned TAGGR for a minuscule amount of ICP - screwing up the established market price (rather than the auction being cancelled due to lack of competition). Have I understood this correctly?

Reply:
Yes, the bids are kept for the next auction. The behaviour you describe is possible if no interest in minted tokens exist from other users. But in this case, the described trick is still possible by anybody to do, it's just a little bit easier.

At the same time, if this happens, no one cares about the token and it is the real market price, right?

Note that the minting from the price will still be capped, so it's just the user who had this bid will get a lot of TAGGR for free (and it would be deserved), but nothing catastrophic should happen.

WDYT?

Reply:
> if this happens, no one cares about the token

Maybe, but the surest way to maximise the amount of TAGGR minted would be for everyone to agree to bid the lowest amount of ICP for the highest amount of TAGGR. You would not necessarily loose out even if your bid was not the first to be submitted. What would be more important is that you *purchase* super cheap TAGGR afterwards (before the cap is reached). I could imagine peer pressure and transparency causing the community to potentially cooperate in this way.

The users that stand to loose out are those that already own lots of TAGGR (inflation due to maximal minting). This would motivate/require such users to place more reasonable bids even though they don't necessarily want any more TAGGR. Due to bid transparency, I could see this leading to more of a 'them and use' mentality (causing division between new and established users - the have and the have nots)

Reply:
I feel like it would be good to have a safety measure, such as querying the [Exchange Rate Canister](https://dashboard.internetcomputer.org/canister/uf6dk-hyaaa-aaaaq-qaaaq-cai) (or another approach for establishing the external market rate for TAGGR) and make sure that the internal make rate (from auctioning) is never less than some arbitrary percentage of the external market rate (say... 90%). This allows for anywhere between 0% and 10% saving by purchasing from the TAGGR platform directly.

This still delivers on generating revenue for the platform, while allowing users to benefit from somewhat cheaper TAGGR, while also protecting against significant price manipulation due to some sort of abuse of the auction.

Reply:
> would be for everyone to agree to bid the lowest amount of ICP for the highest amount of TAGGR

I personally plan to set a weekly timer to check for arbitrage opportunities. Additionally, I'll set a catch-order on BEACON for TAGGR say 25% below the market price. This order can sit forever there. Once filled, again: arbitrage. This alone will make the arbitraging very easy for everyone. I think I won't be alone doing this: this is literally free money, after all. So I can't really imagine how the attack you described can ever work... It assumes, people don't need free money 🤷‍♂️

> is never less than some arbitrary percentage of the external market rate 

But having an "external" price ratio kills the whole purpose of the auction - being autonomous, because now we need to fetch it from somewhere. There should be no knowledge of external market rate: there are simply no good enough decentralized external markets right now. If there would be, we wouldn't need the auction.

Reply:
My feeling was that the point of the auction was to establish a *fair* price that would tend to be slightly lower than the going market rate, providing a reason for users to purchase their TAGGR from the platform (which mints it) rather than from external exchanges (thereby generating more revenue for TAGGR).

I like the idea of avoiding external dependencies, but it just feels to me like getting users to internally establish the market rate that TAGGR will be minted by incentivises those users to try and make it as cheap as possible, so that they can purchase more of it more cheaply. As far as I can tell, the only thing stopping this from happening would be a *tragedy of the commons* type of situation (which is usually referenced as a bad thing), whereby two users notice that all the bids are offering extremely low amounts of ICP for high amounts of TAGGR, so one of them decides to swoop in and bid for all the auctioned TAGGR at a more reasonable rate, but then the other one does the same (say this happens a few times, so the bids become more reasonable). But now because of those two users' actions, the established market rate that TAGGR can be minted by has shot up. Sure, one of them got 15 TAGGR minted to them as a result of the auction, but it they'd just played the cooperative game that everyone else was trying to play, the internal market rate for TAGGR would have been crazy slow, allowing them to purchase significantly more TAGGR much more cheaply after the auction closed.

Given that the established market rate isn't used for selling TAGGR with this release (other than the auction), it wouldn't currently be an issue anyway. But it's something I can imagine being a problem whenever that release comes out. If feels like users stand to gain more by cooperatively gaming the auction, than by competing in the auction - simply because after the auction has closed then would be able to buy TAGGR based on the results of the auction

Reply:
Couldn't we actually fix it as following: after the minting price is established, every user can decide to convert the collected rewards to TAGGR at this price or to ICP? This way, no matter what happens to the auction price, users have the last word wrt. to their own minting.

But yes, I would address this in the next step or so.

Reply:
> Why not make the auction the only means of aquiring TAGGR from the platform? This should increase interest in the auction, and therefore make it more competitive.
> What do you think?

Hm, this is an interesting idea, actually. We need to discuss this on Taggr, with everyone!

> increase it or decrease it automatically on the next auction based on whether or not the previous auction was competitive enough

This is another good idea! However I still think it's not possible that there will be no sell out, because everyone can instantly dump the newly minted Taggr on a market. So it doesn't matter if people collude or not, it take one arbitrageur to do this trade.

Anyway, these discussions are very interesting, so let's merge this PR and continue on Taggr with the new observations?